enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,754 @@
+
+[^7]: CCX collect data that includes the MachineSet `providerSpec` for every cluster that is opted into telemetry and is using the MachineSet feature.
+We can therefore ask them to extract the MachineSets from the data and provide us with enough detail to be able to compare, within each cluster,
+if there are discrepancies that cannot be resolved by our existing planned logic for conversion.
+
+##### Machine Synchronisation
+
+Machine synchronisation will happen in much the same way as in MachineSets.
+A Machine API Machine will map to a Machine and an InfrastructureMachine in Cluster API.
+
+If a Machine is owned by a MachineSet, then the existing InfrastructureTemplate from the MachineSet can be used to create the InfrastructureMachine.
+If the Machine is standalone, the template can be generated in the same way as described in the MachineSet section.
+
+When Cluster API Machines are authoritative, if a Machine API Machine exists for the Cluster API Machine,
+the controller will synchronise the Machine and InfrastructureMachine content onto the Machine API Machine.
+Since the Machien API Machine represents a superset of the Cluster API resources (for supported features), there should be no errors converting in this direction.
+
+
+##### MachineHealthCheck Synchronisation
+
+The fields on a Cluster API MachineHealthCheck form a superset of the Machine API MachineHealthCheck.
+The additional fields are either optional or can be determined easily by the controller (eg. ClusterName).
+Conversion of MachineHealthChecks should be relatively straightforward because of this.
+
+When Cluster API MachineHealthChecks are authoritative, and the optional features of a Cluster API MachineHealthCheck are being leveraged, this will not be synced onto the Machine API MachineHealthCheck.
+This restriction is known and will be documented.
+Additional features provided by Cluster API will not be ported to Machine API as part of this effort.
+
+##### ControlPlaneMachineSet Synchronisation
+
+ControlPlaneMachineSets are not present in Cluster API.
+However, the ControlPlaneMachineSet was designed in such a way that we can extend the `template`` section (which is a discriminated union) to support Cluster API.
+
+A separate enhancement will be created for the support of Cluster API in the ControlPlaneMachineSet, however, for the sake of conversion,
+we expect the conversion to happen much in the same way as a regular MachineSet.
+
+ControlPlaneMachineSets (additionally to MachineSets) have a concept of failure domains.
+This is also a concept in Cluster API.
+When converting ControlPlaneMachineSets to Cluster API, we will ensure that the appropriate failure domain configuration is configured inside the Cluster API InfrastructureCluster,
+and that the failure domains from the ControlPlaneMachineSet Machine API configuration are reflected arcoss.
```

Comment:
Note that in Cluster API a failure domain is a string reference to something understood by the InfrastructureProvider, whereas in CPMS a failure domain is a complete object. This gets complicated when failure domains are modified. CPMS handles this gracefully, but in Cluster API a failure domain *may not be modified*. No controller will notice that, e.g. a failure domain reference by a MachineTemplate has been modified and trigger a rollout. The result will just be inconsistent state. The solution to this is not to modify them, and to instead create new failure domains with a new name.

This may mean creating failure domains based on the hash of their contents as described for MachineSets, but note that previously-created failure domain should likely not be removed.